### ISSUE TITLE
Proposal: expose downstream protocol and stream id as opt-in response headers

### ISSUE THREAD

**Issue by goaway** [envoyproxy/envoy#4092]
*Title*: Expose downstream protocol and stream id as opt-in response headers

*Description*:
We'd like this feature primarily to use as additional context in analytics data. While clients technically have this information already, few publicly-shipped HTTP stacks expose it, and none are consistent.

I have an implementation that exposes this from the http codecs and injects the headers in the http connection manager. @mattklein123  suggested header injection could also be done within the codec itself.

Open questions:
Most important: how do folks feel about this?
Where should the code live?
For http/1.* should we return a notional stream id (e.g. set it as 1 initially and increment if the connection is re-used)?

I feel confident about tackling the implementation, but wanted to get feedback before going further.

**Comment by mattklein123** [comment]
As long as this is opt-in on a config basis I don't have an objection to this. I guess loosely my preference would be to expose this from the codec at the interface layer, and for HTTP/1 just use an incrementing "stream" ID which would still give you info about where on the connection the request was started. So I think how you have likely coded it makes sense to me w/ the config at the HCM layer.

@alyssawilk what is your thinking here?

**Comment by goaway** [comment]
Incidentally, I have the headers tentatively named:
x-envoy-downstream-protocol
x-envoy-downstream-stream-id

Open to other options on that front as well, though.

**Comment by alyssawilk** [comment]
We pass both around internally so this totally passes the 2+ users general purpose usefulness test :-)    +1 to  incremental stream id for HTTP/1 - for network latency analysis it's important to know the request was on a pre-warmed connection.  

My question is how many of these reasonable headers we're going to have.  Internally we've got client protocol, number of requests on a connection, initial server vip, client certs, sever cert selected, initial connection security, ciphers, negotiation mechanism, rtt, srtt, bandwith, etc, etc, etc. for something like 30 fields total, most of which I think will be generally purpose useful and we'll add many of them upstream   Doing each of these as 
x-envoy-client-rtt: #, 
x-envoy-client-srtt: # 
x-envoy-client-cert-hash: [hash]
etc 
ends up adding enough bytes to add up to multiple rtts on non-warm connections.  To avoid this, we have a general purpose header the "frontline info header" where individual routes can opt in or out of specific fields via a mechanism similar to user defined headers.  So you end up with one
x-envoy-frontline-info-header: srtt=#;rtt=#;request=#;cert=hash;
and then there's utilities to pull the token you want from the one general purpose header.  It is a bit more work for upstream consumers to have to parse the tokens, but it's been well worth it to us.  I think we should consider if we want such a thing for Envoy - Google will definitely do so internally even if we have to one-off munge a bunch of upstream headers on the way out so we don't need upstream, it, but if other @envoyproxy/maintainers think it's worth the complexity vs bandwidth trade-off I'd be happy to have it upstream and start it here.

**Comment by goaway** [comment]
Thanks for the feedback, @mattklein123 and @alyssawilk !
@alyssawilk, you raise a good point about the other possible fields - in truth I had already been considering measuring rtts from envoy to the client at some point. A couple thoughts:

First, it seems like we could support both a "packed" format as well as an individual-header format fairly easily. All the plumbing to get at the information would be the same and configuration options could dictate which (if any) we inject. This appeals to me because it lets me defer on biting the complexity bullet of the packed format for now, but it seems totally reasonable that we support that in the not too distant future.

Second, I know we used protobufs for a "packed" tracing header format at one point. For a packed frontline-info header proto jumps out at me as appealing both because it possibly simplifies the work for clients to extract information from the header, and also because it gives us an external schema. The trade-off is of course that it _does_ require a consumer of the header to speak proto and isn't human readable like the format you suggest above.

I don't have terribly strong feelings about this. All of the above seem workable.

**Comment by htuch** [comment]
I think based on previous discussion with @PiotrSikora, we're been persuaded against using proto/JSON for new headers that we're defining if we want these to be somewhat standard. I'm a proto partisan, but I buy the argument that if we want any headers we define to have wider adoption, we can't be dictating proto.

**Comment by mattklein123** [comment]
Yeah I think I would shy away from proto/JSON here and do a basic format. I agree on designing a packed header with all of this information.

Additionally, I would prefer that we configure the output of this header using the "extra debug" information configuration that @junr03 has been designing and is going to start working on. @goaway can you sync with @junr03 next week when he is back and add this to that document? I think that document also needs to go out for wider community review. Thank you!

**Comment by junr03** [comment]
Catching up on this now that I am back.

1. I'll follow up with @goaway to add this to the doc.
2. @alyssawilk is @vilobhmm the primary owner of the `Reverse Proxy Response Propagation to downstreams` doc you sent me? In my doc we mentioned that the work can be separate, but I want to make sure we are not duplicating effort, so I want to sync before we start working on this.
3. @mattklein123, other than resolving point 2 above (and adding that to my doc), I am happy to share with the wider community. Once I get a chance to sync with folks at google I will open up an issue to share the doc. Thoughts?

**Comment by mattklein123** [comment]
Yes please share the other doc whenever we can. Also, feel free to close/consolidate issues that are related to this topic and potentially duplicates.

**Comment by alyssawilk** [comment]
summarizing discussions at the community meeting, we're lean towards using http parameterized lists
https://tools.ietf.org/html/draft-ietf-httpbis-header-structure-06#section-3.3
as fairly readable, semi-colon delimited and standards compliant.

**Comment by junr03** [comment]
re: fitting this into the richer response information design I have been working on.

@goaway and I met, and we agree that this feature fits in nicely into the proposed framework. However, given that I have not shared the document for community comment (and internally at Lyft it is unclear to me when we are going to schedule this work), it seems too vague to me to block adding this feature on that work.

So what we propose is the following: add this feature as an http conn man config option, and then once the richer response information framework is in place we can fold this feature into it (with a deprecation cycle of course).

**Comment by alyssawilk** [comment]
Cool.  @junr03 what do you think of making the config for these 2 new options into their own sub-proto in the hcm?  I think both it'd be nice to keep the frontline-info-header options together, and it might make it easier to migrate when we have the rich info framework.

**Comment by PiotrSikora** [comment]
@alyssawilk yeah, which would make sense, but maybe we're biased towards the edge proxy case. The issue description mentions response headers, and there is already `X-Forwarded-Proto` in Envoy, which `x-envoy-downstream-protocol` in request headers would duplicate, so I want to make sure that we're all talking about the same thing, and not just assume that we do.

**Comment by goaway** [comment]
@PiotrSikora It's my understanding that X-Forwarded-Proto generally is only used to distinguish between TLS and plaintext connections (i.e. either http or https), whereas x-envoy-downstream-protocol is meant to provide the the HTTP-compatible protocol used for the downstream connection (i.e. http/1, http/2, quic, etc.). I am also specifically proposing a response header, though this doesn't preclude support as a forward-propagated request header.

**Comment by mattklein123** [comment]
> Cool. @junr03 what do you think of making the config for these 2 new options into their own sub-proto in the hcm? I think both it'd be nice to keep the frontline-info-header options together, and it might make it easier to migrate when we have the rich info framework.

+1

> I am also specifically proposing a response header, though this doesn't preclude support as a forward-propagated request header.

We are going to use a single header with delimiters, right?

**Comment by cmluciano** [comment]
@junr03 @htuch @alyssawilk should we coalesce this with #4283 ?

**Comment by goaway** [comment]
Sorry for the delay/inactivity - I still am working on this as a low priority change, but since we reached a general consensus here, I don't mind closing the issue or coalescing.
